### ISSUE TITLE
Untangling a line in openTSNE vs FIt-SNE

### ISSUE THREAD

**Issue by dkobak** [pavlin-policar/openTSNE#106]
I've been experimenting with extreme early exaggeration that should approximate Laplacian eigenmaps as shown in https://epubs.siam.org/doi/abs/10.1137/18M1216134. My simple toy example (straight line in 3D) works fine in FIt-SNE, but fails in openTSNE, and I struggle to understand why.

Here is a reproducible example:

```
n = 10000
X = np.zeros((n,3))
X[:,0] = np.arange(n)

Z1 = fast_tsne(X, seed=40, learning_rate=1, early_exag_coeff=n/10, stop_early_exag_iter=2000, max_iter=2000)
Z2 = fast_tsne(X, seed=41, learning_rate=1, early_exag_coeff=n/10, stop_early_exag_iter=2000, max_iter=2000)
Z3 = fast_tsne(X, seed=42, learning_rate=1, early_exag_coeff=n/10, stop_early_exag_iter=2000, max_iter=2000)

Z4 = TSNE(n_jobs=-1, initialization='random', random_state=40, learning_rate=1, early_exaggeration=n/10, 
           early_exaggeration_iter=2000, n_iter=0).fit(X)
Z5 = TSNE(n_jobs=-1, initialization='random', random_state=41, learning_rate=1, early_exaggeration=n/10, 
           early_exaggeration_iter=2000, n_iter=0).fit(X)
Z6 = TSNE(n_jobs=-1, initialization='random', random_state=42, learning_rate=1, early_exaggeration=n/10, 
           early_exaggeration_iter=2000, n_iter=0).fit(X)

plt.figure(figsize=(9,3))
for i,Z in enumerate([Z1,Z2,Z3], 1):
    plt.subplot(1,3,i)
    plt.scatter(Z[:,0], Z[:,1], s=1, c=np.arange(n))
sns.despine()
plt.tight_layout()
plt.savefig('line_fitsne.png')

plt.figure(figsize=(9,3))
for i,Z in enumerate([Z4,Z5,Z6], 1):
    plt.subplot(1,3,i)
    plt.scatter(Z[:,0], Z[:,1], s=1, c=np.arange(n))
sns.despine()
plt.tight_layout()
plt.savefig('line_opentsne.png')
```

I set learning rate to 1 and exaggeration coefficient to n/10, as in the paper above. I do early exaggeration for 2000 iterations and nothing else. Other parameters are default. FIt-SNE successfully unwraps the line for every random seed, as it should.

![line_fitsne](https://user-images.githubusercontent.com/8970231/73140762-c137bf80-407c-11ea-9b35-936af9f857b6.png)

But openTSNE fails to unwrap the line for every random seed:

![line_opentsne](https://user-images.githubusercontent.com/8970231/73140763-cac12780-407c-11ea-9c85-95f646ad7778.png)

It does not look like it's due to convergence: I tried running it for another 1000 iterations (so 3000 in total) and it looks basically the same.
It does not look like it's due to initialization either: I tried manually initializing with random initializations with std=0.0001, and it still does not succeed. 

Weird. I wanted to use openTSNE to make animations of how the line gets untangled with extreme exaggeration, but it only seems to work in FIt-SNE :-(

**Comment by dkobak** [comment]
Hmm. Seems like it *is* a convergence issue: if I run it for 20000 (!) iterations, then two out of three do unwrap:

![line_opentsne_20k](https://user-images.githubusercontent.com/8970231/73143759-3f588e00-409e-11ea-82ee-5515e69b398c.png)

Not sure why openTSNE converges so much slower in this particular test case, even though all optimisation parameters should be the same.

**Comment by pavlin-policar** [comment]
Hmm, that's strange. I do think it's initialization specific, because picking some other random seeds, openTSNE is able to unwrap what line. Seeds 2, 5, and 6 all result in an untangled line.

Using the same random seed in FIt-SNE and openTSNE is probably not going to actually use the same init because I pass this through to numpy, while FIt-SNE just seeds the random number generator with it. I would be surprised if the internals were exactly the same, so the initializations are probably different. I don't have FIt-SNE set up at the moment, so could you check if there is a difference if you actually pass the initialization to both methods? You can use [`openTSNE.initialization.random`](https://github.com/pavlin-policar/openTSNE/blob/master/openTSNE/initialization.py#L6-L30) for initialization, or modify it to have 0.0001 std.

**Comment by dkobak** [comment]
Fair enough. I modified the code to use exact same initialization, and tried other three random seeds:

```
n = 10000
X = np.zeros((n,3))
X[:,0] = np.arange(n)

Zs_fitsne = []
Zs_opentsne = []
for i in range(3):
    np.random.seed(i)
    init = np.random.randn(n,2)
    init = init / np.std(init, axis=0) * 0.0001

    Z = fast_tsne(X, initialization=init, learning_rate=1, early_exag_coeff=n/10, 
                  stop_early_exag_iter=2000, max_iter=2000)
    Zs_fitsne.append(Z)
    
    Z = TSNE(n_jobs=-1, initialization=init, random_state=40, learning_rate=1, 
             early_exaggeration=n/10, early_exaggeration_iter=2000, n_iter=0).fit(X)
    Zs_opentsne.append(Z)

plt.figure(figsize=(9,6))
for i,Z in enumerate(Zs_fitsne + Zs_opentsne, 1):
    plt.subplot(2,3,i)
    plt.scatter(Z[:,0], Z[:,1], s=1, c=np.arange(n))
sns.despine()
plt.tight_layout()
plt.savefig('line_fitsne_opentsne.png')
```

![line_fitsne_opentsne](https://user-images.githubusercontent.com/8970231/73162557-51fab380-40ee-11ea-94f7-26c41924b7cd.png)

Looks like FIt-SNE also sometimes fails to untangle with these particular parameter values, while openTSNE does fork for some seeds. Overall it still seems like openTSNE is converging slower though, and curiously the spans of the embeddings are quite different (10 times difference), despite the same initialization. This must be due to some subtle differences in the optimisation, and I wonder what they can be. That said, given that this line example is arguably rather pathological and only has theoretical interest, this might not be worth investigating any further...

**Comment by pavlin-policar** [comment]
I know there are very minor differences in the optimization, but I really wouldn't expect them to be this big. I have noticed the thing with the spans as well, and I'll definitely be investigating this in the next couple of days. Even if it is such an edge case, it's worth getting to the bottom of it.

**Comment by dkobak** [comment]
Have you already had a chance to look into this?.. I'm preparing some animations for a talk with this line embedding using openTSNE right now :) Might post on twitter :)

**Comment by pavlin-policar** [comment]
Hey, I did spend a little bit looking into it but didn't really figure out why this was happening. I did find that when initializing the embeddings at 0.0001 variance, the span of the embedding after optimization was very small, which was puzzling. I did, however, determine that the difference is not in the subtle difference in the momentum implementation between FIt-SNE and openTSNE. So, in all, I was left with more questions than answers. I'll look into this again when I can, but we've got a couple of deadlines to hit right now.

I'd definitely be interested in seeing those animations as well. The few ones I did when implementing this were pretty cool to look at :)

**Comment by dkobak** [comment]
> Hey, I did spend a little bit looking into it but didn't really figure out why this was happening.

I think I might have figured it out. FIt-SNE has `mom_switch_iter=250` by default and this value (currently) does not change when one supplies a non-default `stop_early_exag_iter`. Whereas openTSNE rather sensibly uses initial momentum for the whole early exaggeration phase. Turns out, this is what makes FIt-SNE converge faster in this case!

Here is a demonstration:

```
n = 10000
X = np.zeros((n,3))
X[:,0] = np.arange(n)

np.random.seed(0)
init = np.random.randn(n,2)
init = init / np.std(init, axis=0) * 0.0001

n_iter = 500

Z1 = fast_tsne(X, initialization=init, learning_rate=1, early_exag_coeff=n/10, 
              stop_early_exag_iter=n_iter, max_iter=n_iter)

Z2 = fast_tsne(X, initialization=init, learning_rate=1, early_exag_coeff=n/10, 
              stop_early_exag_iter=n_iter, max_iter=n_iter, mom_switch_iter=n_iter)
  
Z3 = TSNE(n_jobs=-1, initialization=init, random_state=40, learning_rate=1, 
        early_exaggeration=n/10, early_exaggeration_iter=n_iter, n_iter=0).fit(X)


plt.figure(figsize=(9,3))
for i,Z in enumerate([Z1,Z2,Z3], 1):
    plt.subplot(1,3,i)
    plt.scatter(Z[:,0], Z[:,1], s=1, c=np.arange(n))
    plt.xlim([np.min(Z), np.max(Z)])
    plt.ylim([np.min(Z), np.max(Z)])
sns.despine()
plt.tight_layout()
```

![tmp](https://user-images.githubusercontent.com/8970231/76808246-6a7e7480-67e7-11ea-8b08-355ae35e7de7.png)

First panel: FIt-SNE with 500 early exaggeration iterations and momentum switching at 250. Second: FIt-SNE with 500 ee iterations and momentum not switching. Third: openTSNE. The second = the third, so case closed.

I think what openTSNE does makes more sense so we can close this issue.

I am wondering if we should change it in openTSNE, but this is such an edge case that I think it does not really matter.

**Comment by pavlin-policar** [comment]
Wow, great job figuring this out!

> I am wondering if we should change it in openTSNE, but this is such an edge case that I think it does not really matter.

No, what openTSNE does makes sense, and this feels more like a bug in FIt-SNE, which should be fixed. While the momentum switch is not necessarily tied to the early exaggeration phase, this is not what I would at all expect (but maybe that's just me and George has his own explanation).

**Comment by dkobak** [comment]
> I am wondering if we should change it in openTSNE

Sorry -- I meant "FIt-SNE" here. It's getting late :)

So yes -- I am wondering if we should change it in FIt-SNE, which would be a very easy fix in wrappers. It does converge faster as it currently is though ;-)

**Comment by dkobak** [comment]
FYI @pavlin-policar see the discussion here https://github.com/KlugerLab/FIt-SNE/issues/94
